packages/core/src/sdk/product/usePageProductsQuery.ts

```diff
@@ -77,42 +77,32 @@ export const useProductsPrefetch = (page: number | null) => {
 }
 
 function updatesProductsPerPageRef(
-  productsPerPageRef: MutableRefObject<ProductsPerPage[]>,
+  productsPerPage: ProductsPerPage[] = [],
   page: number,
   data: ClientProductsQueryQuery
 ) {
-  const productsPerPage = productsPerPageRef.current
-  const currentProductsPerPage = { page, data }
-
-  if (productsPerPage.length === 0) {
-    productsPerPage.push(currentProductsPerPage)
-    return { productsPerPage, currentProductsPerPage }
-  }
+  const currentPagedProducts = { page, data }
 
   const index = productsPerPage.findIndex((item) => item.page === page)
-  const lastItem = productsPerPage[productsPerPage.length - 1]
-  const shouldReplaceItem = index !== -1
-  const shouldAddNext = page > lastItem.page
+  const shouldReplacePage = index !== -1
 
-  if (shouldReplaceItem) {
-    productsPerPage.splice(index, 1, currentProductsPerPage)
-  } else if (shouldAddNext) {
-    productsPerPage.push(currentProductsPerPage)
+  if (shouldReplacePage) {
+    productsPerPage[index] = currentPagedProducts
```

Comment:
Here you are mutating the original array. The same thing would be something like this, without mutating:

```suggestion
    const newProductsPerPage = [...currentPagedProducts.slice(0, index), currentPagedProducts, ...currentPagedProducts.slice(index +1, currentPagedProducts.length)]
```

It is more complicated, but doesn't change the original array reference and avoids errors that can come from that.

Reply:
If you want to mutate it, then make sure to create a new array before making the changes:

```suggestion
    const newProductsPerPage = [...productsPerPage]
    newProductsPerPage[index] = currentPagedProducts
```

Reply:
We addressed this with pair programming :)